### DISCUSSION TITLE
Python virtualenvs/Poetry

### DISCUSSION THREAD

**Discussion by tiagoantao** [zed-industries/zed#16533] [Feature Requests]
Hi,

I have started to use Zed and I am finding an ergonomics problem with Python (Poetrry/virtualenvs).

When I have a project, Zed doesn't automatically find the virtual env / poetry install for the project. Which means that the editor is not able to resolve dependencies for packages used in the project.

I have found a couple of answers here, but they require adding a configuration file for pyright - which is very ugly compared to all other editors...

... It so happens that other editors don't require that (and adding a configuration file for a pyright environment would cause problems for other collaborators).

For example with vim I do:

```shell
poetry shell
vim blah.py # all fine here
```

Because vim is run inside the virtual environment the problem is automatically solved. The impression that I have is that other editors are able to solved this problem in whatever way.

Any ideas on how to solve this elegantly (i.e. this is automaticallly solved without adding pyright configuration directives)?

Thanks

**Comment by hongquan** [comment]
This is not a direct answer, but, when adding pyright configuration, you should use `pyrightconfig.json` file and make it ignored by Git, to let other collaborators use their own content.

**Reply by tiagoantao** [reply]
I get that and it works (thanks!), but with other editors this step is not needed. Either because you are inside the enviroment (e.g. with vim) or because they have ways to resolve it

**Reply by hongquan** [reply]
Agree. This should be supported by Zed, because which Python venv is active also affects other features/tools, not just PyRight.

**Comment by pjv** [comment]
I do not use poetry so this is not really an answer in that context and I agree 100% that it would be great to make Zed’s detection of python virtual environments more comprehensive, hands-off, and bullet-proof.

In the mean time, what I’ve done is configure things in my setup (where I use pyenv and direnv to create python projects) so that my virtual environment python install (regardless of what python version a given venv is using) will always be located under `.venv` in the project’s directory.

And then I have the following in my zed `settings.json`:

 ````
  "lsp": {
    "basedpyright": {
        "python": {
          "pythonPath": ".venv/bin/python"
        }
      }
    },
    "pyright": {
      "settings": {
        "python": {
          "pythonPath": ".venv/bin/python"
        }
      }
    },
…
````

In this way, these python LSP’s find the venv in every project without requiring a `pyrightconfig.json` in that project’s directory.

**Comment by hanxuanliang** [comment]
https://www.reddit.com/r/ZedEditor/comments/1cth023/cant_get_zed_to_recognize_any_python_libraries/

It seems this approach can solve the current problem, but Zed is still not smart enough.

**Reply by chumpalump** [reply]
I had to tweak the linked recipe, then it worked perfectly for me.
```
d=$(poetry env info -p); printf '{\n  "venvPath": "%s",\n  "venv": "%s"\n}\n' "$(dirname "$d")" "$(basename "$d")" > pyrightconfig.json
```

**Reply by vshlapakov** [reply]
Thanks, that helped, a similar snippet for Fish shell:
```
set -l d (poetry env info -p); printf '{\n  "venvPath": "%s",\n  "venv": "%s"\n}\n' (dirname "$d") (basename "$d") > pyrightconfig.json
```

**Comment by trieungoctam** [comment]
In pyproject.toml:

```toml
[tool.pyright]
venvPath = "."
venv = ".venv"
```
pyright will look for the virtual environment in the project. Next, we need to configure poetry to create the virtual environment in the project:

```bash
poetry config virtualenvs.in-project true
```

**Reply by hongquan** [reply]
Note that, not all developers want to keep virtual env in the same project folder, and adding this configuration to _pyproject.toml_, which is tracked by Git, will stop teammates from using their own virtual environment.

**Comment by danielward27** [comment]
I'm running into this issue with the debugger, if you want to use the "play" button for tests, then I don't think there is a way to specify the environment used?

**Comment by kenkunz** [comment]
The support documented [here](https://zed.dev/docs/languages/python#pyright-virtual-environments) doesn't work for our setup. Poetry stores virtual environments in a centralized cache directory (e.g., `~/Library/Caches/pypoetry/virtualenvs/` on macOS) rather than in the project directory, so Zed's `detect_venv` feature doesn't find them.

With the latest release (`0.204.1`) that makes `basedpyright` the default LSP for Python, I no longer need to manually set `pythonPath` in my local project settings… which is an improvement. However, I still need to manually activate my virtual environment in terminals:

```
poetry env activate
# above command echos a `source` command with correct path
source /Users/username/Library/Caches/pypoetry/virtualenvs/my-project-ABC123XY-py3.11/bin/activate
```

### Work around ###

I added a vibe-coded (thanks Claude 😄) `.zsh` startup script (using oh my zsh, so added ` ~/.oh-my-zsh/custom/poetry-auto-activate.zsh`):

```zsh
# Poetry auto-activation for Zed terminals only
# File: ~/.oh-my-zsh/custom/poetry-auto-activate.zsh

auto_poetry_shell() {
    # Only run in Zed terminals
    if [[ "$TERM_PROGRAM" != "zed" ]]; then
        return
    fi

    # Activate virtual environment
    if [[ -f "pyproject.toml" ]] && [[ -z "$VIRTUAL_ENV" ]]; then
        if command -v poetry >/dev/null 2>&1; then
            local venv_path=$(poetry env info --path 2>/dev/null)
            if [[ -n "$venv_path" ]] && [[ -d "$venv_path" ]]; then
                source "$venv_path/bin/activate"
                echo "🐍 Activated Poetry venv: $(basename $venv_path)"
            fi
        fi
    fi
}

# Hook into directory changes and shell startup
autoload -U add-zsh-hook
add-zsh-hook chpwd auto_poetry_shell
auto_poetry_shell  # Run on shell startup

# Optional: Add a quick alias for manual activation
# alias poe='source $(poetry env info --path)/bin/activate'
```

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
